File: backend/src/subscriptions/nudge.ts
import { Entitlement, EntitlementTier } from './entitlement';

/** Which nudge strip a client should render. NONE = render nothing. */
export enum NudgeBanner {
  NONE = 'none',
  PRO_EXPIRING = 'pro_expiring',
  JUST_EXPIRED = 'just_expired',
  FREE_COUNTER = 'free_counter',
}

export interface NudgeState {
  tier: EntitlementTier;
  usageToday: number;
  dailyLimit: number;
  expiresAt: string | null;
  banner: NudgeBanner;
}

const DAY_MS = 86_400_000;
const PRO_EXPIRING_DAYS = 3;
const JUST_EXPIRED_DAYS = 7;

/**
 * Pure mapping from entitlement → banner. `lastExpiredAt` is the
 * updated_at of the user's most recent EXPIRED row (null if never).
 * `now` is injected for testability.
 */
export function deriveBanner(
  entitlement: Entitlement,
  _usageToday: number,
  lastExpiredAt: Date | null,
  now: Date,
): NudgeBanner {
  if (entitlement.tier === EntitlementTier.PRO) {
    if (entitlement.expiresAt) {
      const daysLeft = (entitlement.expiresAt.getTime() - now.getTime()) / DAY_MS;
      if (daysLeft <= PRO_EXPIRING_DAYS) return NudgeBanner.PRO_EXPIRING;
    }
    return NudgeBanner.NONE;
  }
  if (lastExpiredAt) {
    const daysSince = (now.getTime() - lastExpiredAt.getTime()) / DAY_MS;
    if (daysSince <= JUST_EXPIRED_DAYS) return NudgeBanner.JUST_EXPIRED;
  }
  return NudgeBanner.FREE_COUNTER;
}

File: backend-rewrite-go/internal/subscription/nudge.go
package subscription

import "time"

// NudgeBanner mirrors the NestJS NudgeBanner enum string values.
type NudgeBanner string

const (
	BannerNone        NudgeBanner = "none"
	BannerProExpiring NudgeBanner = "pro_expiring"
	BannerJustExpired NudgeBanner = "just_expired"
	BannerFreeCounter NudgeBanner = "free_counter"
)

// Banner-window thresholds from nudge.ts. Comparisons inclusive (<=),
// arithmetic on unrounded day-fraction floats (matches the TS).
const (
	dayMS           = float64(86_400_000)
	proExpiringDays = 3.0
	justExpiredDays = 7.0
	// FreeDailyLimit is the free-tier dailyLimit fallback when no plan applies.
	FreeDailyLimit = 10
)

// DeriveBanner ports subscriptions/nudge.ts deriveBanner. usageToday is
// intentionally absent — the TS signature ignores it. tier is "pro" or
// "free"; expiresAt is the active sub's expiry (pro path); lastExpiredAt
// is the updated_at of the most-recent expired row (free path).
func DeriveBanner(tier string, expiresAt, lastExpiredAt *time.Time, now time.Time) NudgeBanner {
	if tier == "pro" {
		if expiresAt != nil {
			daysLeft := float64(expiresAt.Sub(now).Milliseconds()) / dayMS
			if daysLeft <= proExpiringDays {
				return BannerProExpiring
			}
		}
		return BannerNone
	}
	if lastExpiredAt != nil {
		daysSince := float64(now.Sub(*lastExpiredAt).Milliseconds()) / dayMS
		if daysSince <= justExpiredDays {
			return BannerJustExpired
		}
	}
	return BannerFreeCounter
}
